File: controllers/adminController.js
const pool = require('../config/db');

// GET /api/admin/dashboard
exports.getDashboardStats = async (req, res) => {
  try {
    const [totalUsers] = await pool.query('SELECT COUNT(*) as count FROM users');
    const [activeFarmers] = await pool.query('SELECT COUNT(*) as count FROM users WHERE role = "farmer" AND status = "active"');
    const [activeBuyers] = await pool.query('SELECT COUNT(*) as count FROM users WHERE role = "buyer" AND status = "active"');
    const [totalProducts] = await pool.query('SELECT COUNT(*) as count FROM products');
    const [totalOrders] = await pool.query('SELECT COUNT(*) as count FROM orders');
    const [revenue] = await pool.query('SELECT COALESCE(SUM(total_amount), 0) as total FROM orders WHERE payment_status = "paid"');

    // Recent users
    const [recentUsers] = await pool.query(
      'SELECT id, name, email, role, status, created_at FROM users ORDER BY created_at DESC LIMIT 5'
    );

    // Recent orders
    const [recentOrders] = await pool.query(
      `SELECT o.*, u.name as buyer_name
       FROM orders o
       JOIN users u ON o.buyer_id = u.id
       ORDER BY o.created_at DESC LIMIT 5`
    );

    res.json({
      stats: {
        totalUsers: totalUsers[0].count,
        activeFarmers: activeFarmers[0].count,
        activeBuyers: activeBuyers[0].count,
        totalProducts: totalProducts[0].count,
        totalOrders: totalOrders[0].count,
        revenue: revenue[0].total
      },
      recentUsers,
      recentOrders
    });
  } catch (err) {
    console.error('Admin dashboard error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/admin/users
exports.getAllUsers = async (req, res) => {
  try {
    const { search, role, status } = req.query;

    let query = 'SELECT id, name, email, role, phone, location, business_name, status, is_verified, created_at FROM users WHERE 1=1';
    const params = [];

    if (search) {
      query += ' AND (name LIKE ? OR email LIKE ?)';
      params.push(`%${search}%`, `%${search}%`);
    }
    if (role) { query += ' AND role = ?'; params.push(role); }
    if (status) { query += ' AND status = ?'; params.push(status); }

    query += ' ORDER BY created_at DESC';

    const [users] = await pool.query(query, params);
    res.json({ users });
  } catch (err) {
    console.error('Get all users error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/admin/farmers
exports.getAllFarmers = async (req, res) => {
  try {
    const { search } = req.query;
    let query = `
      SELECT u.id, u.name, u.email, u.phone, u.location, u.business_name, u.status, u.created_at,
        (SELECT COUNT(*) FROM products p WHERE p.farmer_id = u.id) as product_count,
        (SELECT COUNT(*) FROM order_items oi WHERE oi.farmer_id = u.id) as order_count,
        (SELECT COALESCE(SUM(oi.price_at_purchase * oi.quantity), 0) FROM order_items oi JOIN orders o ON oi.order_id = o.id WHERE oi.farmer_id = u.id AND o.payment_status = 'paid') as total_earnings
      FROM users u WHERE u.role = 'farmer'
    `;
    const params = [];
    if (search) {
      query += ' AND (u.name LIKE ? OR u.email LIKE ?)';
      params.push(`%${search}%`, `%${search}%`);
    }
    query += ' ORDER BY u.created_at DESC';

    const [farmers] = await pool.query(query, params);
    res.json({ farmers });
  } catch (err) {
    console.error('Get all farmers error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/admin/buyers
exports.getAllBuyers = async (req, res) => {
  try {
    const { search } = req.query;
    let query = `
      SELECT u.id, u.name, u.email, u.phone, u.location, u.business_name, u.status, u.created_at,
        (SELECT COUNT(*) FROM orders o WHERE o.buyer_id = u.id) as order_count,
        (SELECT COALESCE(SUM(o.total_amount), 0) FROM orders o WHERE o.buyer_id = u.id AND o.payment_status = 'paid') as total_spent
      FROM users u WHERE u.role = 'buyer'
    `;
    const params = [];
    if (search) {
      query += ' AND (u.name LIKE ? OR u.email LIKE ?)';
      params.push(`%${search}%`, `%${search}%`);
    }
    query += ' ORDER BY u.created_at DESC';

    const [buyers] = await pool.query(query, params);
    res.json({ buyers });
  } catch (err) {
    console.error('Get all buyers error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/admin/transactions
exports.getTransactions = async (req, res) => {
  try {
    const { search, status } = req.query;
    let query = `
      SELECT o.*, u.name as buyer_name, u.email as buyer_email,
        (SELECT COUNT(*) FROM order_items oi WHERE oi.order_id = o.id) as item_count
      FROM orders o
      JOIN users u ON o.buyer_id = u.id
      WHERE 1=1
    `;
    const params = [];
    if (search) {
      query += ' AND (u.name LIKE ? OR u.email LIKE ? OR o.id = ?)';
      params.push(`%${search}%`, `%${search}%`, search);
    }
    if (status && status !== 'All Status') {
      query += ' AND o.status = ?';
      params.push(status.toLowerCase());
    }
    query += ' ORDER BY o.created_at DESC';

    const [transactions] = await pool.query(query, params);
    res.json({ transactions });
  } catch (err) {
    console.error('Get transactions error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/admin/analytics
exports.getAnalytics = async (req, res) => {
  try {
    // Monthly order trends (last 6 months)
    const [monthlyOrders] = await pool.query(`
      SELECT DATE_FORMAT(created_at, '%Y-%m') as month,
        COUNT(*) as orders,
        COALESCE(SUM(total_amount), 0) as revenue
      FROM orders
      WHERE created_at >= DATE_SUB(NOW(), INTERVAL 6 MONTH)
      GROUP BY DATE_FORMAT(created_at, '%Y-%m')
      ORDER BY month ASC
    `);

    // Category breakdown
    const [categoryBreakdown] = await pool.query(`
      SELECT p.category, COUNT(*) as product_count,
        COALESCE(SUM(oi.quantity * oi.price_at_purchase), 0) as revenue
      FROM products p
      LEFT JOIN order_items oi ON p.id = oi.product_id
      GROUP BY p.category
    `);

    // User growth (last 6 months)
    const [userGrowth] = await pool.query(`
      SELECT DATE_FORMAT(created_at, '%Y-%m') as month,
        SUM(CASE WHEN role = 'farmer' THEN 1 ELSE 0 END) as farmers,
        SUM(CASE WHEN role = 'buyer' THEN 1 ELSE 0 END) as buyers
      FROM users
      WHERE created_at >= DATE_SUB(NOW(), INTERVAL 6 MONTH)
      GROUP BY DATE_FORMAT(created_at, '%Y-%m')
      ORDER BY month ASC
    `);

    // Top products
    const [topProducts] = await pool.query(`
      SELECT p.name, p.category, COUNT(oi.id) as order_count,
        COALESCE(SUM(oi.quantity * oi.price_at_purchase), 0) as revenue
      FROM products p
      JOIN order_items oi ON p.id = oi.product_id
      GROUP BY p.id
      ORDER BY order_count DESC
      LIMIT 5
    `);

    res.json({
      monthlyOrders,
      categoryBreakdown,
      userGrowth,
      topProducts
    });
  } catch (err) {
    console.error('Analytics error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// PUT /api/admin/users/:id/status
exports.updateUserStatus = async (req, res) => {
  try {
    const { status } = req.body;
    await pool.query('UPDATE users SET status = ? WHERE id = ?', [status, req.params.id]);
    res.json({ message: 'User status updated.' });
  } catch (err) {
    console.error('Update user status error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

File: backend-spring/src/main/java/com/agrivalue/controller/AdminController.java
package com.agrivalue.controller;

import com.agrivalue.entity.Order;
import com.agrivalue.entity.User;
import com.agrivalue.repository.*;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/admin")
public class AdminController {

    private final UserRepository userRepo;
    private final ProductRepository productRepo;
    private final OrderRepository orderRepo;
    private final OrderItemRepository orderItemRepo;

    public AdminController(UserRepository userRepo, ProductRepository productRepo,
                           OrderRepository orderRepo, OrderItemRepository orderItemRepo) {
        this.userRepo = userRepo;
        this.productRepo = productRepo;
        this.orderRepo = orderRepo;
        this.orderItemRepo = orderItemRepo;
    }

    @GetMapping("/dashboard")
    public ResponseEntity<?> getDashboardStats() {
        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("totalUsers", userRepo.count());
        stats.put("activeFarmers", userRepo.countByRoleAndStatus(User.Role.farmer, User.Status.active));
        stats.put("activeBuyers", userRepo.countByRoleAndStatus(User.Role.buyer, User.Status.active));
        stats.put("totalProducts", productRepo.count());
        stats.put("totalOrders", orderRepo.count());
        stats.put("revenue", orderRepo.getTotalRevenue());

        List<Map<String, Object>> recentUsers = userRepo.findAll(
                org.springframework.data.domain.PageRequest.of(0, 5,
                        org.springframework.data.domain.Sort.by("createdAt").descending())
        ).getContent().stream().map(u -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", u.getId()); m.put("name", u.getName()); m.put("email", u.getEmail());
            m.put("role", u.getRole().name()); m.put("status", u.getStatus().name()); m.put("created_at", u.getCreatedAt());
            return m;
        }).collect(Collectors.toList());

        List<Map<String, Object>> recentOrders = orderRepo.findAllByOrderByCreatedAtDesc().stream().limit(5).map(o -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", o.getId()); m.put("total_amount", o.getTotalAmount());
            m.put("status", o.getStatus().name()); m.put("created_at", o.getCreatedAt());
            userRepo.findById(o.getBuyerId()).ifPresent(u -> m.put("buyer_name", u.getName()));
            return m;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("stats", stats, "recentUsers", recentUsers, "recentOrders", recentOrders));
    }

    @GetMapping("/users")
    public ResponseEntity<?> getAllUsers(@RequestParam(required = false) String search,
                                         @RequestParam(required = false) String role,
                                         @RequestParam(required = false) String status) {
        List<User> users = userRepo.findAll();
        if (search != null && !search.isEmpty()) {
            String s = search.toLowerCase();
            users = users.stream().filter(u -> u.getName().toLowerCase().contains(s) || u.getEmail().toLowerCase().contains(s)).collect(Collectors.toList());
        }
        if (role != null && !role.isEmpty()) {
            users = users.stream().filter(u -> u.getRole().name().equals(role)).collect(Collectors.toList());
        }
        if (status != null && !status.isEmpty()) {
            users = users.stream().filter(u -> u.getStatus().name().equals(status)).collect(Collectors.toList());
        }

        List<Map<String, Object>> result = users.stream().map(u -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", u.getId()); m.put("name", u.getName()); m.put("email", u.getEmail());
            m.put("role", u.getRole().name()); m.put("phone", u.getPhone()); m.put("location", u.getLocation());
            m.put("business_name", u.getBusinessName()); m.put("status", u.getStatus().name());
            m.put("is_verified", u.getIsVerified()); m.put("created_at", u.getCreatedAt());
            return m;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("users", result));
    }

    @GetMapping("/farmers")
    public ResponseEntity<?> getAllFarmers(@RequestParam(required = false, defaultValue = "") String search) {
        List<User> farmers = search.isEmpty() ? userRepo.findByRole(User.Role.farmer) :
                userRepo.findByRoleAndSearch(User.Role.farmer, search);

        List<Map<String, Object>> result = farmers.stream().map(f -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", f.getId()); m.put("name", f.getName()); m.put("email", f.getEmail());
            m.put("phone", f.getPhone()); m.put("location", f.getLocation());
            m.put("business_name", f.getBusinessName()); m.put("status", f.getStatus().name());
            m.put("created_at", f.getCreatedAt());
            m.put("product_count", productRepo.countByFarmerId(f.getId()));
            m.put("order_count", orderItemRepo.countByFarmerId(f.getId()));
            m.put("total_earnings", orderItemRepo.getTotalEarningsByFarmerId(f.getId()));
            return m;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("farmers", result));
    }

    @GetMapping("/buyers")
    public ResponseEntity<?> getAllBuyers(@RequestParam(required = false, defaultValue = "") String search) {
        List<User> buyers = search.isEmpty() ? userRepo.findByRole(User.Role.buyer) :
                userRepo.findByRoleAndSearch(User.Role.buyer, search);

        List<Map<String, Object>> result = buyers.stream().map(b -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", b.getId()); m.put("name", b.getName()); m.put("email", b.getEmail());
            m.put("phone", b.getPhone()); m.put("location", b.getLocation());
            m.put("business_name", b.getBusinessName()); m.put("status", b.getStatus().name());
            m.put("created_at", b.getCreatedAt());
            long orderCount = orderRepo.findByBuyerIdOrderByCreatedAtDesc(b.getId()).size();
            BigDecimal totalSpent = orderRepo.findByBuyerIdOrderByCreatedAtDesc(b.getId()).stream()
                    .filter(o -> o.getPaymentStatus() == Order.PaymentStatus.paid)
                    .map(Order::getTotalAmount).reduce(BigDecimal.ZERO, BigDecimal::add);
            m.put("order_count", orderCount);
            m.put("total_spent", totalSpent);
            return m;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("buyers", result));
    }

    @GetMapping("/transactions")
    public ResponseEntity<?> getTransactions(@RequestParam(required = false) String search,
                                              @RequestParam(required = false) String status) {
        List<Order> orders = orderRepo.findAllByOrderByCreatedAtDesc();
        if (status != null && !status.isEmpty() && !status.equalsIgnoreCase("All Status")) {
            orders = orders.stream().filter(o -> o.getStatus().name().equalsIgnoreCase(status)).collect(Collectors.toList());
        }
        if (search != null && !search.isEmpty()) {
            String s = search.toLowerCase();
            orders = orders.stream().filter(o -> {
                Optional<User> u = userRepo.findById(o.getBuyerId());
                return o.getId().toString().contains(s) ||
                        (u.isPresent() && (u.get().getName().toLowerCase().contains(s) || u.get().getEmail().toLowerCase().contains(s)));
            }).collect(Collectors.toList());
        }

        List<Map<String, Object>> result = orders.stream().map(o -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", o.getId()); m.put("total_amount", o.getTotalAmount());
            m.put("status", o.getStatus().name()); m.put("payment_status", o.getPaymentStatus().name());
            m.put("created_at", o.getCreatedAt());
            m.put("item_count", orderItemRepo.findByOrderId(o.getId()).size());
            userRepo.findById(o.getBuyerId()).ifPresent(u -> { m.put("buyer_name", u.getName()); m.put("buyer_email", u.getEmail()); });
            return m;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("transactions", result));
    }

    @PutMapping("/users/{id}/status")
    public ResponseEntity<?> updateUserStatus(@PathVariable Integer id, @RequestBody Map<String, String> body) {
        userRepo.findById(id).ifPresent(u -> {
            u.setStatus(User.Status.valueOf(body.get("status")));
            userRepo.save(u);
        });
        return ResponseEntity.ok(Map.of("message", "User status updated."));
    }
}
